src/manifest/mod.rs

```diff
@@ -214,15 +214,34 @@ pub struct ManifestAndUnsedKeys {
 
 impl CrateData {
     /// Reads all metadata for the crate whose manifest is inside the directory
-    /// specified by `path`.
-    pub fn new(crate_path: &Path, out_name: Option<String>) -> Result<CrateData, Error> {
-        let manifest_path = crate_path.join("Cargo.toml");
+    /// specified by `path`. If `recurse` is true then walk up the path until
+    /// a manifest file is found, or return an error if one cannot be found.
+    pub fn new(
+        crate_path: &Path,
+        recurse: bool,
+        out_name: Option<String>,
+    ) -> Result<CrateData, Error> {
+        let full_path = crate_path.canonicalize()?;
+        let full_path = full_path.as_path();
+        let mut manifest_path = crate_path.join("Cargo.toml");
+        if recurse {
```

Comment:
I think this logic should be in `set_crate_path` (which should probably be renamed *get* instead of *set*) so that we don't have to deal with a `bool` part in a returned tuple nor a `bool` parameter. It would be cleaner without that, IMO.

Reply:
That was the part that I thought was quite clunky but I was worried about the scope of the refactoring. As it turns of it was the opposite of what I thought - it is the smaller refactoring and is a lot cleaner and obvious. Thanks!